File: jqtype-ts/src/stream.ts
import { JType, type JType as JTypeT } from "./types.js";

export type Cardinality =
  | "Zero"
  | "One"
  | "ZeroOrOne"
  | "OneOrMore"
  | "ZeroOrMore";

export const Cardinality = {
  asStr(c: Cardinality): string {
    switch (c) {
      case "Zero": return "zero";
      case "One": return "one";
      case "ZeroOrOne": return "zero_or_one";
      case "OneOrMore": return "one_or_more";
      case "ZeroOrMore": return "zero_or_more";
    }
  },

  join(a: Cardinality, b: Cardinality): Cardinality {
    if (a === "Zero") return b;
    if (b === "Zero") return a;
    if (a === "One" && b === "One") return "OneOrMore";
    if (
      (a === "One" && b === "ZeroOrOne") ||
      (a === "ZeroOrOne" && b === "One") ||
      (a === "ZeroOrOne" && b === "ZeroOrOne")
    ) {
      return "ZeroOrMore";
    }
    return "ZeroOrMore";
  },

  alternative(a: Cardinality, b: Cardinality): Cardinality {
    if (a === "Zero" && b === "Zero") return "Zero";
    if (
      (a === "Zero" && b === "One") ||
      (a === "One" && b === "Zero") ||
      (a === "Zero" && b === "ZeroOrOne") ||
      (a === "ZeroOrOne" && b === "Zero")
    ) {
      return "ZeroOrOne";
    }
    if (
      (a === "Zero" && b === "OneOrMore") ||
      (a === "OneOrMore" && b === "Zero") ||
      (a === "Zero" && b === "ZeroOrMore") ||
      (a === "ZeroOrMore" && b === "Zero")
    ) {
      return "ZeroOrMore";
    }
    if (a === "One" && b === "One") return "One";
    if (
      (a === "One" && b === "ZeroOrOne") ||
      (a === "ZeroOrOne" && b === "One") ||
      (a === "ZeroOrOne" && b === "ZeroOrOne")
    ) {
      return "ZeroOrOne";
    }
    if (
      (a === "One" && b === "OneOrMore") ||
      (a === "OneOrMore" && b === "One") ||
      (a === "OneOrMore" && b === "OneOrMore")
    ) {
      return "OneOrMore";
    }
    return "ZeroOrMore";
  },

  fitsCount(c: Cardinality, count: number): boolean {
    switch (c) {
      case "Zero": return count === 0;
      case "One": return count === 1;
      case "ZeroOrOne": return count <= 1;
      case "OneOrMore": return count >= 1;
      case "ZeroOrMore": return true;
    }
  },

  compose(outer: Cardinality, inner: Cardinality): Cardinality {
    if (outer === "Zero" || inner === "Zero") return "Zero";
    if (outer === "One") return inner;
    if (outer === "ZeroOrOne") {
      if (inner === "One") return "ZeroOrOne";
      if (inner === "ZeroOrOne") return "ZeroOrOne";
      return "ZeroOrMore";
    }
    if (outer === "OneOrMore") {
      if (inner === "One") return "OneOrMore";
      if (inner === "OneOrMore") return "OneOrMore";
      return "ZeroOrMore";
    }
    return "ZeroOrMore";
  },
};

export interface StreamType {
  item: JTypeT;
  card: Cardinality;
}

export const StreamType = {
  new(item: JTypeT, card: Cardinality): StreamType {
    if (card === "Zero") return { item: "Never", card };
    return { item, card };
  },

  one(item: JTypeT): StreamType {
    return StreamType.new(item, "One");
  },

  zero(): StreamType {
    return StreamType.new("Never", "Zero");
  },

  zeroOrOne(item: JTypeT): StreamType {
    return StreamType.new(item, "ZeroOrOne");
  },

  zeroOrMore(item: JTypeT): StreamType {
    return StreamType.new(item, "ZeroOrMore");
  },

  join(a: StreamType, b: StreamType): StreamType {
    return StreamType.new(
      JType.union([a.item, b.item]),
      Cardinality.join(a.card, b.card),
    );
  },

  joinAlternative(a: StreamType, b: StreamType): StreamType {
    return StreamType.new(
      JType.union([a.item, b.item]),
      Cardinality.alternative(a.card, b.card),
    );
  },

  toCompactString(s: StreamType): string {
    if (s.card === "One") return JType.toCompactString(s.item);
    return `Stream<${JType.toCompactString(s.item)}, ${s.card}>`;
  },

  fitsOutputs(
    s: StreamType,
    outputs: unknown[],
    itemCheck: (value: unknown, ty: JTypeT) => boolean,
  ): boolean {
    if (!Cardinality.fitsCount(s.card, outputs.length)) return false;
    return outputs.every((value) => itemCheck(value, s.item));
  },
};

File: jqtype-rs/crates/jqtype-core/src/stream.rs
use serde::{Deserialize, Serialize};

use crate::types::JType;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum Cardinality {
    Zero,
    One,
    ZeroOrOne,
    OneOrMore,
    ZeroOrMore,
}

impl Cardinality {
    pub fn as_str(&self) -> &'static str {
        match self {
            Cardinality::Zero => "zero",
            Cardinality::One => "one",
            Cardinality::ZeroOrOne => "zero_or_one",
            Cardinality::OneOrMore => "one_or_more",
            Cardinality::ZeroOrMore => "zero_or_more",
        }
    }

    pub fn join(self, other: Self) -> Self {
        use Cardinality::*;
        match (self, other) {
            (Zero, x) | (x, Zero) => x,
            (One, One) => OneOrMore,
            (One, ZeroOrOne) | (ZeroOrOne, One) | (ZeroOrOne, ZeroOrOne) => ZeroOrMore,
            (OneOrMore, _) | (_, OneOrMore) | (ZeroOrMore, _) | (_, ZeroOrMore) => ZeroOrMore,
        }
    }

    pub fn alternative(self, other: Self) -> Self {
        use Cardinality::*;
        match (self, other) {
            (Zero, Zero) => Zero,
            (Zero, One) | (One, Zero) | (Zero, ZeroOrOne) | (ZeroOrOne, Zero) => ZeroOrOne,
            (Zero, OneOrMore) | (OneOrMore, Zero) => ZeroOrMore,
            (Zero, ZeroOrMore) | (ZeroOrMore, Zero) => ZeroOrMore,
            (One, One) => One,
            (One, ZeroOrOne) | (ZeroOrOne, One) | (ZeroOrOne, ZeroOrOne) => ZeroOrOne,
            (One, OneOrMore) | (OneOrMore, One) | (OneOrMore, OneOrMore) => OneOrMore,
            (One, ZeroOrMore)
            | (ZeroOrMore, One)
            | (ZeroOrOne, OneOrMore)
            | (OneOrMore, ZeroOrOne)
            | (ZeroOrOne, ZeroOrMore)
            | (ZeroOrMore, ZeroOrOne)
            | (OneOrMore, ZeroOrMore)
            | (ZeroOrMore, OneOrMore)
            | (ZeroOrMore, ZeroOrMore) => ZeroOrMore,
        }
    }

    /// Returns `true` when `count` is a valid number of values for this
    /// cardinality (e.g. `One` accepts exactly one, `OneOrMore` accepts any
    /// non-zero count).
    pub fn fits_count(&self, count: usize) -> bool {
        use Cardinality::*;
        match self {
            Zero => count == 0,
            One => count == 1,
            ZeroOrOne => count <= 1,
            OneOrMore => count >= 1,
            ZeroOrMore => true,
        }
    }

    pub fn compose(self, inner: Self) -> Self {
        use Cardinality::*;
        match (self, inner) {
            (Zero, _) | (_, Zero) => Zero,
            (One, c) => c,
            (ZeroOrOne, One) => ZeroOrOne,
            (ZeroOrOne, ZeroOrOne) => ZeroOrOne,
            (ZeroOrOne, OneOrMore | ZeroOrMore) => ZeroOrMore,
            (OneOrMore, One) => OneOrMore,
            (OneOrMore, ZeroOrOne | ZeroOrMore) => ZeroOrMore,
            (OneOrMore, OneOrMore) => OneOrMore,
            (ZeroOrMore, _) => ZeroOrMore,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct StreamType {
    pub item: JType,
    pub card: Cardinality,
}

impl StreamType {
    pub fn new(item: JType, card: Cardinality) -> Self {
        if matches!(card, Cardinality::Zero) {
            Self {
                item: JType::Never,
                card,
            }
        } else {
            Self { item, card }
        }
    }

    pub fn one(item: JType) -> Self {
        Self::new(item, Cardinality::One)
    }

    pub fn zero() -> Self {
        Self::new(JType::Never, Cardinality::Zero)
    }

    pub fn zero_or_one(item: JType) -> Self {
        Self::new(item, Cardinality::ZeroOrOne)
    }

    pub fn zero_or_more(item: JType) -> Self {
        Self::new(item, Cardinality::ZeroOrMore)
    }

    pub fn join(self, other: Self) -> Self {
        Self::new(
            JType::union([self.item, other.item]),
            self.card.join(other.card),
        )
    }

    pub fn join_alternative(self, other: Self) -> Self {
        Self::new(
            JType::union([self.item, other.item]),
            self.card.alternative(other.card),
        )
    }

    pub fn to_compact_string(&self) -> String {
        if matches!(self.card, Cardinality::One) {
            self.item.to_compact_string()
        } else {
            format!("Stream<{}, {:?}>", self.item.to_compact_string(), self.card)
        }
    }

    /// Returns `true` when `outputs` is a possible concrete output of a
    /// filter with this stream type. Used by the compatibility harness.
    pub fn fits_outputs(
        &self,
        outputs: &[serde_json::Value],
        item_check: impl Fn(&serde_json::Value, &JType) -> bool,
    ) -> bool {
        if !self.card.fits_count(outputs.len()) {
            return false;
        }
        outputs.iter().all(|value| item_check(value, &self.item))
    }
}
